evm/src/test/java/org/hyperledger/besu/evm/operations/SelfDestructOperationTest.java

```diff
@@ -0,0 +1,185 @@
+            .build();
+    messageFrame.pushStackItem(Bytes.fromHexString(beneficiary));
+
+    when(mutableAccount.getBalance()).thenReturn(Wei.fromHexString(balanceHex));
+    when(mutableAccount.getAddress()).thenReturn(contractAddress, beneficiaryAddress);
+    when(worldUpdater.getAccount(any())).thenReturn(account);
+    when(worldUpdater.get(any())).thenReturn(account);
+    when(worldUpdater.getOrCreate(any())).thenReturn(account);
+  }
+
+  public static Object[][] params() {
+    return new Object[][] {
+      {
+        "0x00112233445566778899aabbccddeeff11223344",
+        "0x1234567890abcdef1234567890abcdef12345678",
+        true,
+        "0x1234567890"
+      },
+      {
+        "0x00112233445566778899aabbccddeeff11223344",
+        "0x1234567890abcdef1234567890abcdef12345678",
+        false,
+        "0x1234567890"
+      },
+      {
+        "0x00112233445566778899aabbccddeeff11223344",
+        "0x00112233445566778899aabbccddeeff11223344",
+        true,
+        "0x1234567890"
+      },
+      {
+        "0x1234567890abcdef1234567890abcdef12345678",
+        "0x1234567890abcdef1234567890abcdef12345678",
+        false,
+        "0x1234567890"
+      },
+    };
+  }
+
+  @ParameterizedTest
+  @MethodSource("params")
+  void checkContractDeletionFrontier(
+      final String contract,
+      final String beneficiary,
+      final boolean ignoredNewAccount,
+      final String balanceHex) {
+
+    setUp(contract, beneficiary, balanceHex);
+    final Operation.OperationResult operationResult = operation.execute(messageFrame, evm);
+    assertThat(operationResult).isNotNull();
+    assertThat(messageFrame.getSelfDestructs()).contains(Address.fromHexString(contract));
+
+    verify(account).getBalance();
+    verify(account).isEmpty();
+    if (!contract.equals(beneficiary)) {
+      verify(mutableAccount).incrementBalance(Wei.fromHexString(balanceHex));
+    }
+    verify(mutableAccount).setBalance(Wei.ZERO);
+
+    verifyNoMoreInteractions(worldUpdater, account, mutableAccount, evm, newMutableAccount);
+  }
+
+  @ParameterizedTest
+  @MethodSource("params")
+  void checkContractDeletionEIP6780(
+      final String contract,
+      final String beneficiary,
+      final boolean newAccount,
+      final String balanceHex) {
+
+    setUp(contract, beneficiary, balanceHex);
+    when(mutableAccount.isNewAccount()).thenReturn(newAccount);
+    final Operation.OperationResult operationResult = newOperation.execute(messageFrame, evm);
+    assertThat(operationResult).isNotNull();
+    if (newAccount) {
+      assertThat(messageFrame.getSelfDestructs()).contains(Address.fromHexString(contract));
+    } else {
+      assertThat(messageFrame.getSelfDestructs()).isEmpty();
+    }
+
+    verify(account).getBalance();
+    verify(account).isEmpty();
+    if (!contract.equals(beneficiary)) {
+      verify(mutableAccount).incrementBalance(Wei.fromHexString(balanceHex));
+    }
+    verify(mutableAccount).setBalance(Wei.ZERO);
```

Comment:
What does it mean for the contract to be equal to the beneficiary?

Reply:
Contract address is the same as the beneficiary address for the self destruct.  This is reflecting implementation details where we don't increment the account being self-destructed if it's the same account.

Reply:
Balance is indeed ether is lost.  State tests confirm this is the expected outcome.

```
./ethereum/evmtool/build/install/evmtool/bin/evm state-test ./ethereum/referencetests/src/reference-test/external-resources/GeneralStateTests/stSystemOperationsTest/suicideAddress.json --json --json-alloc 
-------8<--------
{"pc":0,"op":48,"gas":"0xef038","gasCost":"0x2","memSize":0,"stack":[],"depth":1,"refund":0,"opName":"ADDRESS"}
{"pc":1,"op":96,"gas":"0xef036","gasCost":"0x3","memSize":0,"stack":["0x95e7baea6a6c7c4c2dfeb977efac326af552d87"],"depth":1,"refund":0,"opName":"PUSH1"}
{"pc":3,"op":85,"gas":"0xef033","gasCost":"0x5654","memSize":0,"stack":["0x95e7baea6a6c7c4c2dfeb977efac326af552d87","0x0"],"depth":1,"refund":0,"opName":"SSTORE"}
{"pc":4,"op":48,"gas":"0xe99df","gasCost":"0x2","memSize":0,"stack":[],"depth":1,"refund":0,"opName":"ADDRESS"}
{"pc":5,"op":255,"gas":"0xe99dd","gasCost":"0x1388","memSize":0,"stack":["0x95e7baea6a6c7c4c2dfeb977efac326af552d87"],"depth":1,"refund":0,"opName":"SELFDESTRUCT"}
{
 "0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b": {
  "balance": "0xde0b6b3a75b2232",
  "nonce": "0x1"
 },
}
{"output":"","gasUsed":"0xbbeb","time":3454083,"Mgps":"13.928","test":"suicideAddress","fork":"Shanghai","d":0,"g":0,"v":0,"postHash":"0x74d6e975dfc354eca790f5a3bee573e0019bb92967c9131ddc2dc622f297e266","postLogsHash":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","pass":true}
```

`0x95e7baea6a6c7c4c2dfeb977efac326af552d87` was the executing contract, and had balance